src/vs/base/common/comparers.ts

```diff
@@ -6,30 +6,168 @@
+	}
+
+	return 0;
 }
 
-const FileNameMatch = /^(.*?)(\.([^.]*))?$/;
+/** Compares filenames by name case, then by name, then by extension. Sorts uppercase names before lowercase. */
+export function compareFileNamesUpper(one: string | null, other: string | null) {
+	const [oneName, oneExtension] = extractNameAndExtension(one);
+	const [otherName, otherExtension] = extractNameAndExtension(other);
+	const collator = intlFileNameCollator.getValue().collator;
+	let result;
+
+	// Check for case differences in names.
+	result = compareCaseUpperFirst(oneName, otherName);
+	if (result !== 0) {
+		return result;
+	}
+
+	// Extension case should not be considered when grouping files by case.
+	// Do not check here for case differences in extensions.
+
+	// Check for name differences.
+	result = collator.compare(oneName, otherName);
+	if (result !== 0) {
+		return result;
+	}
+
+	// Check for extension differences, including possible case differences.
+	return collator.compare(oneExtension, otherExtension);
+}
+
+/** Compares filenames by name case, then by name, then by extension. Sorts lowercase names before uppercase. */
+export function compareFileNamesLower(one: string | null, other: string | null) {
+	const [oneName, oneExtension] = extractNameAndExtension(one);
+	const [otherName, otherExtension] = extractNameAndExtension(other);
+	const collator = intlFileNameCollator.getValue().collator;
+	let result;
 
+	// Check for case differences in names.
+	result = compareCaseLowerFirst(oneName, otherName);
+	if (result !== 0) {
+		return result;
+	}
+
+	// Extension case should not be considered when grouping files by case.
+	// Do not check here for case differences in extensions.
+
+	// Check for name differences.
+	result = collator.compare(oneName, otherName);
+	if (result !== 0) {
+		return result;
+	}
+
+	// Check for extension differences, including possible case differences.
+	return collator.compare(oneExtension, otherExtension);
+}
+
+/** Compares filenames by name, then by extension. */
+export function compareFileNamesMixed(one: string | null, other: string | null) {
+	const [oneName, oneExtension] = extractNameAndExtension(one);
+	const [otherName, otherExtension] = extractNameAndExtension(other);
+	const collator = intlFileNameCollator.getValue().collator;
+	let result;
+
+	// Check for name differences.
+	result = collator.compare(oneName, otherName);
+	if (result !== 0) {
+		return result;
+	}
+
+	// Check for extension differences.
+	return collator.compare(oneExtension, otherExtension);
+}
+
+/** Compares filenames by unicode value, not differentiating between names and extensions. */
+export function compareFileNamesUnicode(one: string | null, other: string | null) {
+	one = one || '';
+	other = other || '';
+
+	// Simply compare both strings. No name vs extension awareness.
+	if (one === other) {
+		return 0;
+	}
+
+	return one < other ? -1 : 1;
+}
+
+/** Compares filenames by name unicode value, then extension unicode value. */
```

Comment:
This function doesn't seem to be used anywhere except in extHostDocumentData.test.ts. Could it possibly be removed? Or does it need to stick around?